instrumentation/nocode/src/main/java/com/splunk/opentelemetry/instrumentation/nocode/NocodeInstrumentation.java

```diff
@@ -0,0 +1,93 @@
+import static io.opentelemetry.javaagent.bootstrap.Java8BytecodeBridge.currentContext;
+import static net.bytebuddy.matcher.ElementMatchers.hasSuperType;
+import static net.bytebuddy.matcher.ElementMatchers.named;
+
+import com.splunk.opentelemetry.javaagent.bootstrap.nocode.NocodeRules;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.context.Scope;
+import io.opentelemetry.instrumentation.api.incubator.semconv.util.ClassAndMethod;
+import io.opentelemetry.javaagent.extension.instrumentation.TypeInstrumentation;
+import io.opentelemetry.javaagent.extension.instrumentation.TypeTransformer;
+import java.lang.reflect.Method;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+
+public final class NocodeInstrumentation implements TypeInstrumentation {
+  private final NocodeRules.Rule rule;
+
+  public NocodeInstrumentation(NocodeRules.Rule rule) {
+    this.rule = rule;
+  }
+
+  @Override
+  public ElementMatcher<TypeDescription> typeMatcher() {
+    return hasSuperType(named(rule.className));
+  }
+
+  @Override
+  public void transform(TypeTransformer transformer) {
+    transformer.applyAdviceToMethod(
+        named(rule.methodName), this.getClass().getName() + "$NocodeAdvice");
+  }
+
+  @SuppressWarnings("unused")
+  public static class NocodeAdvice {
+    @Advice.OnMethodEnter(suppress = Throwable.class)
+    public static void onEnter(
+        @Advice.Origin("#t") Class<?> declaringClass,
+        @Advice.Origin("#m") String methodName,
+        @Advice.Local("otelInvocation") NocodeMethodInvocation otelInvocation,
+        @Advice.Local("otelContext") Context context,
+        @Advice.Local("otelScope") Scope scope,
+        @Advice.This Object thiz,
+        @Advice.AllArguments Object[] methodParams) {
+      NocodeRules.Rule rule = NocodeRules.find(declaringClass.getName(), methodName);
+      otelInvocation =
+          new NocodeMethodInvocation(
+              rule, ClassAndMethod.create(declaringClass, methodName), thiz, methodParams);
+      Context parentContext = currentContext();
+
+      if (!instrumentor().shouldStart(parentContext, otelInvocation)) {
+        return;
+      }
+      context = instrumentor().start(parentContext, otelInvocation);
+      scope = context.makeCurrent();
+    }
+
+    @Advice.OnMethodExit(onThrowable = Throwable.class, suppress = Throwable.class)
+    public static void stopSpan(
+        @Advice.Origin Method method,
+        @Advice.Local("otelInvocation") NocodeMethodInvocation otelInvocation,
+        @Advice.Local("otelContext") Context context,
+        @Advice.Local("otelScope") Scope scope,
+        @Advice.Thrown Throwable error) {
+      if (scope == null) {
+        return;
+      }
+      scope.close();
+      // I wonder why the original methods instrumentation had support for modifying the return
```

Comment:
this is because the original instrumentation has special handling for some async types such as `CompletableFuture` where span is ended when the future ends not when the method returns

Reply:
Yes, on rereading that code I see that now.  I'll change the comment to reflect this.